com.unity.netcode.gameobjects/Prototyping/Interpolator/BufferedLinearInterpolator.cs

```diff
@@ -0,0 +1,232 @@
+        {
+            return UseFixedUpdate ? t.FixedTime : t.Time;
+        }
+
+        // todo if I have value 1, 2, 3 and I'm treating 1 to 3, I shouldn't interpolate between 1 and 3, I should interpolate from 1 to 2, then from 2 to 3 to get the best path
+        private void TryConsumeFromBuffer()
+        {
+            int consumedCount = 0;
+            // only consume if we're ready
+            if (RenderTime >= m_EndTimeConsumed.Time)
+            {
+                // buffer is sorted so older (smaller) time values are at the end.
+                for (int i = m_Buffer.Count - 1; i >= 0; i--) // todo stretch: consume ahead if we see we're missing values
+                {
+                    var bufferedValue = m_Buffer[i];
+                    // Consume when ready. This can consume multiple times
+                    if (FixedOrTime(bufferedValue.timeSent) <= ServerTimeBeingHandledForBuffering)
+                    {
+                        if (m_LifetimeConsumedCount == 0)
+                        {
+                            m_StartTimeConsumed = bufferedValue.timeSent;
+                            m_InterpStartValue = bufferedValue.item;
+                        }
+                        else if (consumedCount == 0)
+                        {
+                            m_StartTimeConsumed = m_EndTimeConsumed;
+                            m_InterpStartValue = m_InterpEndValue;
+                        }
+
+                        m_EndTimeConsumed = bufferedValue.timeSent;
+                        m_InterpEndValue = bufferedValue.item;
+                        m_Buffer.RemoveAt(i);
+                        consumedCount++;
+                        m_LifetimeConsumedCount++;
+                    }
+                }
+            }
+        }
+
+        public T Update(float deltaTime)
+        {
+            TryConsumeFromBuffer();
+
+            if (m_LifetimeConsumedCount == 0 && m_Buffer.Count == 0)
+            {
+                throw new InvalidOperationException("trying to update interpolator when no data has been added to it yet");
+            }
+
+            // Interpolation example to understand the math below
+            // 4   4.5      6   6.5
+            // |   |        |   |
+            // A   render   B   Server
+
+            if (m_LifetimeConsumedCount >= 1) // shouldn't interpolate between default values, let's wait to receive data first, should only interpolate between real measurements
+            {
+                double range = FixedOrTime(m_EndTimeConsumed) - FixedOrTime(m_StartTimeConsumed);
+                float t;
+                if (range == 0)
+                {
+                    t = 1;
+                }
+                else
+                {
+                    t = (float) ((RenderTime - FixedOrTime(m_StartTimeConsumed)) / range);
+                }
+
+                if (t > 3) // max extrapolation
+                {
+                    t = 1;
+                }
```

Comment:
huh?

Reply:
if we skip positions because of packet loss, unclamped lerping will extrapolate and continue iterating in that direction. However, we don't want that extrapolation to continue to infinity. If we still don't receive anything and we go beyond 3 ticks, we go back to the last "true" value we received

Reply:
make this behaviour configurable?

Reply:
ok. I'll look into this in a followup PR, there's also an issue with teleportation right now and this, I'll need to come back to it.